crates/rune/src/parse/lexer.rs

```diff
@@ -559,12 +599,72 @@ impl<'a> Lexer<'a> {
+                            self.iter.next();
+                            let (doc, inner) = match self.iter.peek() {
+                                Some('/') => {
+                                    // The character following this must not be another /.
+                                    // If it is, it's probably a separator, not a doc comment.
+                                    if let Some('/') = self.iter.peek2() {
+                                        (false, false)
+                                    } else {
+                                        (true, false)
+                                    }
+                                }
+                                Some('!') => {
+                                    (true, true)
+                                }
+                                _ => {
+                                    (false, false)
+                                }
+                            };
+
                             self.consume_line();
-                            break ast::Kind::Comment;
+                            if doc {
+                                // docstring span drops the first 3 characters (/// or //!)
+                                let span = self.iter.span_to_pos(start);
+                                self.emit_doc_attribute(inner, span, span.clone().trim_start(3));
+                                continue 'outer;
+                            } else {
+                                break ast::Kind::Comment;
+                            }
                         }
                         ('/', '*') => {
+                            self.iter.next();
+                            let (doc, inner) = match self.iter.peek() {
+                                Some('*') => {
+                                    // The character following this must not be another *.
+                                    // If it is, it's probably a separator, not a doc comment.
+                                    if let Some('*') = self.iter.peek2() {
+                                        (false, false)
+                                    } else {
+                                        (true, false)
+                                    }
+                                }
+                                Some('!') => {
+                                    (true, true)
+                                }
+                                _ => {
+                                    (false, false)
+                                }
+                            };
+
                             let term = self.consume_multiline_comment();
-                            break ast::Kind::MultilineComment(term);
+                            if !term {
+                                return Err(ParseError::new(
+                                    self.iter.span_to_pos(start),
+                                    ParseErrorKind::ExpectedMultilineCommentTerm,
```

Comment:
Long term I'd probably prefer if this error was emitted at a later stage, like during parsing. Since then the lexer is more resilient and can be re-used for parsers which are more suitable for live editing where comments are incomplete (future plan).

All though at the moment it makes little difference.